midend/optimizeSelectLabels.cpp

```diff
@@ -0,0 +1,428 @@
+        leftSubRange.first = leafNode->rangeValue.first;
+        leftSubRange.second = value - 1;
+
+        rightSubRange.first = value + 1;
+        rightSubRange.second = leafNode->rangeValue.second;
+
+        big_int leftValidRange = getValidRange(leftSubRange);
+        big_int rightValidRange = getValidRange(rightSubRange);
+
+        if (leftValidRange != -1) {
+            if (leftValidRange != 0)
+                leafNode->left = new SelectLabel(leftValidRange, root->base,
+                                                 root->rangeConstantTypeInfo);
+            else
+                leafNode->left = new SelectLabel(leftSubRange, root->base,
+                                                 root->typeInfo, root->rangeConstantTypeInfo);
+        }
+
+        if (rightValidRange != -1) {
+            if (rightValidRange != 0)
+                leafNode->right = new SelectLabel(rightValidRange, root->base,
+                                                 root->rangeConstantTypeInfo);
+            else
+                leafNode->right = new SelectLabel(rightSubRange, root->base,
+                                                  root->typeInfo, root->rangeConstantTypeInfo);
+        }
+    }
+    return leafNode;
+}
+
+SelectLabel* DoOptimizeSelectLabels::insertValues(SelectLabel* root,
+                                                            std::pair<big_int, big_int> value) {
+    SelectLabel* mainRoot = root;
+    SelectLabel* leafNode = nullptr;
+
+    // Search leaf node where value will be inserted
+    while (mainRoot != nullptr) {
+        leafNode = mainRoot;
+        if (mainRoot->type == SelectLabel::RangeLabel) {
+            // Overlaping two ranges on left side (eg. 1..6 and 4..9 -> 5..9)
+            if ((value.first < mainRoot->rangeValue.first) &&
+                    ((value.second >= mainRoot->rangeValue.first) &&
+                     (value.second < mainRoot->rangeValue.second))) {
+                mainRoot = mainRoot->left;
+            // Overlaping two ranges on right side (eg. 4..9 and 1..6 -> 1..3)
+            } else if ((value.second > mainRoot->rangeValue.second) &&
+                            ((value.first > mainRoot->rangeValue.first) &&
+                                (value.first <= mainRoot->rangeValue.second))) {
+                if (mainRoot->left != nullptr) {
+                    if (mainRoot->left->type == SelectLabel::RangeLabel) {
+                        if (((value.first >= mainRoot->left->rangeValue.first) &&
+                                    (value.first <= mainRoot->left->rangeValue.second)) ||
+                                        ((value.second >= mainRoot->left->rangeValue.first) &&
+                                            (value.second <= mainRoot->left->rangeValue.second)))
+                            mainRoot = mainRoot->left;
+                        else
+                            mainRoot = mainRoot->right;
+                    } else {
```

Comment:
If you handle the else case first you can reduce one level of indentation and make the code more readable:
`if (... != ...) return nullptr;`